pallets/phala/src/stakepool.rs

```diff
@@ -411,52 +422,111 @@ pub mod pallet {
 		NoWhitelistCreated,
 		/// Too long for pool description length
 		ExceedMaxDescriptionLen,
-	}
+		/// Withdraw queue is not empty so that we can't restart mining
+		WithdrawQueueNotEmpty,
+		/// Stakepool's collection_id isn't founded
+		MissingCollectionId,
 
-	#[pallet::hooks]
-	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T>
-	where
-		T: mining::Config<Currency = <T as Config>::Currency>,
-		BalanceOf<T>: FixedPointConvert + Display,
-	{
-		fn on_finalize(_n: T::BlockNumber) {
-			let now = <T as registry::Config>::UnixTime::now()
-				.as_secs()
-				.saturated_into::<u64>();
-			Self::maybe_force_withdraw(now);
-		}
+		InvaildWithdrawSharesAmount,
+
+		AssetAccountNotExist,
 	}
 
 	#[pallet::call]
 	impl<T: Config> Pallet<T>
 	where
-		T: mining::Config<Currency = <T as Config>::Currency>,
-		BalanceOf<T>: FixedPointConvert + Display,
+		BalanceOf<T>: sp_runtime::traits::AtLeast32BitUnsigned + Copy + FixedPointConvert + Display,
+		T: pallet_uniques::Config<CollectionId = CollectionId, ItemId = NftId>,
+		T: pallet_assets::Config<AssetId = u32, Balance = BalanceOf<T>>,
 	{
 		/// Creates a new stake pool
 		#[pallet::weight(0)]
+		#[frame_support::transactional]
 		pub fn create(origin: OriginFor<T>) -> DispatchResult {
 			let owner = ensure_signed(origin)?;
-
-			let pid = PoolCount::<T>::get();
-			StakePools::<T>::insert(
+			let pid = basepool::Pallet::<T>::cosume_new_pid();
+			// TODO(mingxuan): create_collection should return cid
+			let collection_id: CollectionId = pallet_rmrk_core::Pallet::<T>::collection_index();
+			// Create a NFT collection related to the new stake pool
+			let symbol: BoundedVec<u8, <T as pallet_rmrk_core::Config>::CollectionSymbolLimit> =
+				format!("STAKEPOOL-{}", pid)
+					.as_bytes()
+					.to_vec()
+					.try_into()
+					.expect("create a bvec from string should never fail; qed.");
+			pallet_rmrk_core::Pallet::<T>::create_collection(
+				Origin::<T>::Signed(basepool::pallet_id::<T::AccountId>()).into(),
+				Default::default(),
+				None,
+				symbol,
+			)?;
+			basepool::pallet::Pools::<T>::insert(
 				pid,
-				PoolInfo {
-					pid,
-					owner: owner.clone(),
+				PoolProxy::StakePool(StakePool::<T::AccountId, BalanceOf<T>> {
```

Comment:
```suggestion
				PoolProxy::StakePool(StakePool {
```

Have a try. Usually it should be able to infer the correct type.

Reply:
done